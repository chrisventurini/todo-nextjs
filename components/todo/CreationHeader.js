import moment from 'moment';

import AppBar from '@material-ui/core/AppBar';
import Button from '@material-ui/core/Button';
import Toolbar from '@material-ui/core/Toolbar';
import InputBase from '@material-ui/core/InputBase';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles';
import { fade } from '@material-ui/core/styles/colorManipulator';


const styles = theme => ({
    form: {
        flexGrow: 1,
    },
    dateInput: {
        color: 'white',
        marginLeft: '20px'
    },
    todoInput: {
        backgroundColor: fade(theme.palette.common.white, 0.15),
        '&:hover': {
            backgroundColor: fade(theme.palette.common.white, 0.25),
        },
        borderRadius: theme.shape.borderRadius,
        padding: '5px',
        paddingLeft: '10px',
        paddingRight: '10px',
        width: '300px'
    }
});

const CreationHeader = ({todo, classes, onSubmit, onInputChange}) => {

    let dueDate = moment(todo.dueDate).format('YYYY-MM-DD');

    return (
        <AppBar id="header">
            <Toolbar>
                <Typography variant="h6" color="inherit">
                    <form noValidate autoComplete="off" onSubmit={onSubmit}>
                        <InputBase
                            id="to-do-input"
                            className={classes.todoInput}
                            placeholder="Title"
                            name="title"
                            value={todo.title}
                            variant="outlined"
                            onChange={onInputChange}
                        />
                        <TextField
                            id="due-date"
                            className={classes.dateInput}
                            label="Due Date"
                            name="dueDate"
                            value={dueDate}
                            type="date"
                            onChange={onInputChange}
                        />
                        <Button
                            disabled={todo.title ? false : true}
                            type='submit'
                            className={classes.button}>
                            Create Todo
                        </Button>
                    </form>
                </Typography>
            </Toolbar>
        </AppBar>
    )
};

export default withStyles(styles)(CreationHeader);

