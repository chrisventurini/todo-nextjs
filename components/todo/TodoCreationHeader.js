import moment from 'moment';
import propTypes from 'prop-types';

import AppBar from '@material-ui/core/AppBar';
import Button from '@material-ui/core/Button';
import Toolbar from '@material-ui/core/Toolbar';
import InputBase from '@material-ui/core/InputBase';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core/styles';
import { fade } from '@material-ui/core/styles/colorManipulator';


const styles = theme => ({
    button: {
        color: 'black',
        backgroundColor: 'white',
        marginLeft: '20px',

        '&:disabled': {
            opacity: '0.6'
        }
    },
    fieldSet: {
        all: 'unset',

        '&:disabled *': {
            opacity: '0.6'
        }
    },
    form: {
        flexGrow: 1,
    },

    dateInput: {
        color: 'white',
        marginLeft: '20px',

        '& *': {
            color: 'white'
        }
    },
    todoInput: {
        color: 'white',
        backgroundColor: fade(theme.palette.common.white, 0.15),
        '&:hover': {
            backgroundColor: fade(theme.palette.common.white, 0.25),
        },
        borderRadius: theme.shape.borderRadius,
        padding: '5px',
        paddingLeft: '10px',
        paddingRight: '10px',
        width: '300px'
    }
});

const TodoCreationHeader = ({asyncCallsInProgress, todo, classes, onSubmit, onInputChange}) => {

    let dueDate = moment(todo.dueDate).format('YYYY-MM-DD');

    return (
        <AppBar>
            <Toolbar>
                <form noValidate autoComplete="off" onSubmit={onSubmit}>
                    <fieldset className={classes.fieldSet} disabled={asyncCallsInProgress}>
                        <InputBase
                            id="to-do-input"
                            className={classes.todoInput}
                            placeholder="Title"
                            name="title"
                            value={todo.title}
                            variant="outlined"
                            onChange={onInputChange}
                        />
                        <TextField
                            id="due-date"
                            className={classes.dateInput}
                            label="Due Date"
                            name="dueDate"
                            value={dueDate}
                            type="date"
                            onChange={onInputChange}
                        />
                        <Button
                            className={classes.button}
                            disabled={todo.title ? false : true}
                            type='submit'
                            className={classes.button}>
                            Create Todo
                        </Button>
                    </fieldset>
                </form>
            </Toolbar>
        </AppBar>
    )
};

TodoCreationHeader.propTypes = {
    asyncCallsInProgress: propTypes.bool.isRequired
};

export default withStyles(styles)(TodoCreationHeader);

