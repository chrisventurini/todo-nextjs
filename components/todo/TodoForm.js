import Link from 'next/link';
import PropTypes from 'prop-types';
import moment from 'moment';

import Button from '@material-ui/core/Button';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import TextField from '@material-ui/core/TextField';
import withStyles from "@material-ui/core/styles/withStyles";

const styles = {
    todoFieldSet: {
        all: 'unset',

        '&:disabled *': {
            opacity: '0.6'
        }
    },
    todoEdit: {
        marginTop: '55px',
        padding: '10px 25px',
    },

    todoInputs: {
        paddingBottom: '10px',
        width: '100%'
    },

    todoEditCheck: {
        paddingBottom: '10px',
        paddingLeft: '5px',
        width: '100%'
    },

    todoEditControls: {
        marginTop: '10px',
        width: '100%',

        '& button': {
            margin: '0 10px'
        }
    }
};

const TodoForm = ({ asyncCallsInProgress, todo, classes, onDelete, onInputChange, onSubmit}) => {

    let dueDate = moment(todo.dueDate).format('YYYY-MM-DD');

    return (
        <form className={classes.todoEdit} onSubmit={onSubmit}>
            <fieldset className={classes.todoFieldSet} disabled={asyncCallsInProgress}>
                <TextField
                    className={classes.todoInputs}
                    label="Title"
                    name="title"
                    margin="normal"
                    onChange={onInputChange}
                    value={todo.title}
                />
                <FormControlLabel
                    className={classes.todoEditCheck}
                    control={
                        <Checkbox
                            name="completed"
                            onChange={onInputChange}
                            checked={todo.completed}
                            color="primary"
                        />
                    }
                    label="Completed"
                />
                <TextField
                    className={classes.todoInputs}
                    id="due-date"
                    label="Due Date"
                    type="date"
                    name="dueDate"
                    onChange={onInputChange}
                    value={dueDate}
                />
                <TextField
                    className={classes.todoInputs}
                    id="notes"
                    label="Notes"
                    name="notes"
                    onChange={onInputChange}
                    value={todo.notes}
                />
                <div className={classes.todoEditControls}>
                    <Button type="submit" variant="contained" color="primary">
                        Save
                    </Button>
                    <Button onClick={onDelete} variant="contained" color="secondary">
                        X Delete
                    </Button>
                    <Link href="/">
                        <Button variant="contained" color="secondary">
                            Cancel
                        </Button>
                    </Link>
                </div>
            </fieldset>
        </form>
    )
};

TodoForm.propTypes = {
    asyncCallsInProgress: PropTypes.bool.required,
    onDelete: PropTypes.func.required
};

export default withStyles(styles)(TodoForm);

