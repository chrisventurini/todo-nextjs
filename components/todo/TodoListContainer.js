import { connect } from 'react-redux';
import propTypes from 'prop-types';

import todoSorter from '../../services/todoSorter';

import TodoList from './TodoList';


let TodoListContainer = ({todos}) => {
    return (
        <TodoList todos={todos}/>
    )
};

TodoListContainer.propTypes = {
    todos: propTypes.array.isRequired
};

const mapState = state => {
    let { todos } = state;

    if(state.filters && state.filters.completed) {
        todos = todos.filter(todo => !todo.completed);
    }

    todos = todoSorter(todos);

    return { todos }
};

export default connect(mapState)(TodoListContainer);
