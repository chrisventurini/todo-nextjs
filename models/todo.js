module.exports = function (sequelize, DataTypes) {
    let Todo = sequelize.define('todo', {
        id: {
            type:DataTypes.UUID,
            primaryKey: true,
            defaultValue: DataTypes.UUIDV4
        },
        title: DataTypes.STRING,
    });

    return Todo;
};