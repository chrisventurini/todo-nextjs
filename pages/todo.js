import '../styles.scss';

export default () => {
    return (
        <div>This will be a todo page</div>
    )
}
