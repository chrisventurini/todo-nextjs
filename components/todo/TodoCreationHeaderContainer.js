import { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import TodoCreationHeader from '../../components/todo/TodoCreationHeader';

import { mapDispatchToTodoActions } from '../../store/actions/todos/index';

class TodoCreationHeaderContainer extends Component {

    state = {
        title: '',
        dueDate: new Date()
    };

    constructor(props) {
        super(props);

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
    }

    handleInputChange(event) {
        let target = event.target;
        let value = target.value;
        let name = target.name;

        if (target.type === 'date') {
            value = moment(value).format('YYYY-MM-DD');
        }

        this.setState({
            ...this.state,
            [name]: value
        });
    }

    handleSubmit(event) {
        event.preventDefault();
        this.props.actions.todoSave({
            ...this.state
        });

        this.setState({
           title: '',
           dueDate: new Date()
        });
    }

    render() {
        return (
            <TodoCreationHeader
                asyncCallsInProgress={this.props.asyncCalls.inProgress}
                todo={this.state}
                onInputChange={this.handleInputChange}
                onSubmit={this.handleSubmit}
            />
        )
    }
}

const mapState = state => {
    return { asyncCalls: state.asyncCalls };
};

export default connect(mapState, mapDispatchToTodoActions)(TodoCreationHeaderContainer)
