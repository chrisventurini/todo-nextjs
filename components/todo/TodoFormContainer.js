import react, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import TodoForm from './TodoForm';

import { mapDispatchToTodoActions } from '../../store/actions/todos';


class TodoFormContainer extends Component {

    static propTypes = {
        todo: PropTypes.object.isRequired,
        todoId: PropTypes.string.isRequired
    };

    state = {
        completed: false,
        dueDate: new Date(),
        notes: '',
        title: ''
    };

    constructor(props) {
        super(props);


        if(props.todo) {
            this.state = props.todo;
            this.state.notes = props.todo.notes || '';
        }

        this.handleDelete = this.handleDelete.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);

        this.formRef = react.createRef();
    }

    handleDelete(event) {
        event.preventDefault();

        this.props.actions.todoDelete(this.state);
    }

    handleInputChange(event) {
        event.preventDefault();

        let target = event.target,
            value = target.type === 'checkbox' ? target.checked : target.value,
            name = target.name;

        if (event.type === 'date') {
            value = moment(value).format('YYYY-MM-DD');
        }

        this.setState({
            ...this.state,
            [name]: value
        });
    }

    handleSubmit(event) {
        event.preventDefault();

        this.props.actions.todoUpdate({...this.state});
    }

    render () {
        return (
            <TodoForm
                asyncCallsInProgress={this.props.asyncCalls.inProgress}
                ref={this.formRef}
                todo={this.state}
                onDelete={this.handleDelete}
                onSubmit={this.handleSubmit}
                onInputChange={this.handleInputChange}
            />
        )
    }

}

const mapState = (state, ownProps) => {
    let todo = state.todos.collection.find(todo => todo.id === ownProps.todoId);

    return {
        asyncCalls: state.asyncCalls,
        todo
    };
};

export default connect(mapState, mapDispatchToTodoActions)(TodoFormContainer)
