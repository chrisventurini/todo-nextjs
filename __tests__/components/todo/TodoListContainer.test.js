import React from "react";
import { shallow } from "enzyme/build";

import { _mapState, TodoListContainer } from '../../../components/todo/TodoListContainer';

//Mocked imports
jest.mock('../../../services/todoSorter');
import mockTodoSorter from '../../../services/todoSorter';

describe('<TodoListContainer />', () => {
    let SUTWrapper,
        count,
        todos;

    describe('constructing and rendering', () => {

        beforeEach(() => {
            todos = [{}, {}];
            count = todos.length;
            SUTWrapper = shallow(<TodoListContainer todos={todos} count={count} />).dive();
        });

        it('should render correctly', () => {
            expect(SUTWrapper).toMatchSnapshot();
        });

        it('should pass the todos to the TodoList component', () => {
            let foundList = SUTWrapper.first();

            expect(foundList.props()).toHaveProperty('todos', todos);
        });

    });

});

describe('<TodoListContainer /> _mapState', () => {

    describe('when executing', () => {
        let results,
            todos,
            state;

        beforeEach(() => {
            todos = [
                {completed: true},
                {completed: true},
                {completed: false},
                {completed: false}
            ];

            state = {
                filters: {},
                todos: {
                    collection: todos,
                    count: todos.length
                }
            };

            mockTodoSorter.mockImplementation(todos => todos);
        });

        afterAll(() => {
            jest.resetAllMocks();
        });

        describe('with the completed filtered enabled', () => {

            beforeEach(() => {
                state.filters.completed = true;
                results = _mapState(state);
            });

            it('should set the todos collection on the props without the completed todos', () => {
                let expectedTodos = todos.filter(todo => !todo.completed);

                expect(results).toHaveProperty('todos', expectedTodos);
            });

            it('should set the todos count to the props', () => {
                expect(results).toHaveProperty('count', todos.length);
            });

            it('should sort the todos with the todoSorter', () => {
                expect(mockTodoSorter).toHaveBeenCalled();
            });

        });

        describe('with the completed filtered disabled', () => {

            beforeEach(() => {
                state.filters.completed = false;
                results = _mapState(state);
            });

            it('should set the todos collection on the props with the completed todos', () => {
                expect(results).toHaveProperty('todos', todos);
            });

            it('should set the todos count to the props', () => {
                expect(results).toHaveProperty('count', todos.length);
            });

            it('should sort the todos with the todoSorter', () => {
                expect(mockTodoSorter).toHaveBeenCalled();
            });

        });

    });

});
