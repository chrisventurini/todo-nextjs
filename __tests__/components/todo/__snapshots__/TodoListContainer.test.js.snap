// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoListContainer /> constructing and rendering should render correctly 1`] = `
<TodoList
  classes={
    Object {
      "todoList": "TodoList-todoList-1",
    }
  }
  todos={
    Array [
      Object {},
      Object {},
    ]
  }
/>
`;
